pkg/envoy/server.go

```diff
@@ -522,34 +544,64 @@ func getHTTPRule(certManager policy.CertificateManager, h *api.PortRuleHTTP, ns
 				HeaderMatchSpecifier: &envoy_api_v2_route.HeaderMatcher_PresentMatch{PresentMatch: true}})
 		}
 	}
-	for _, hdr := range h.SecretHeaders {
-		// Fetch the secret
-		value := ""
-		var err error
-		if certManager == nil {
-			err = fmt.Errorf("Nil certManager")
-		} else {
-			value, err = certManager.GetSecretString(context.TODO(), hdr.Secret, ns)
+
+	headerMatches := make([]*cilium.HeaderMatch, 0, len(h.HeaderMatches))
+	for _, hdr := range h.HeaderMatches {
+		var mismatch_action cilium.HeaderMatch_MismatchAction
+		switch hdr.Mismatch {
+		case api.MismatchActionLog:
+			mismatch_action = cilium.HeaderMatch_CONTINUE_ON_MISMATCH
+		case api.MismatchActionAdd:
+			mismatch_action = cilium.HeaderMatch_ADD_ON_MISMATCH
+		case api.MismatchActionDelete:
+			mismatch_action = cilium.HeaderMatch_DELETE_ON_MISMATCH
+		case api.MismatchActionReplace:
+			mismatch_action = cilium.HeaderMatch_REPLACE_ON_MISMATCH
+		default:
+			mismatch_action = cilium.HeaderMatch_FAIL_ON_MISMATCH
 		}
+		// Fetch the secret
+		value, err := getSecretString(certManager, hdr, ns)
 		if err != nil {
-			log.WithError(err).Warning("Failed fetching Secret, header match will fail")
+			log.WithError(err).Warning("Failed fetching K8s Secret, header match will fail")
 			// Envoy treats an empty exact match value as matching ANY value; adding
 			// InvertMatch: true here will cause this rule to NEVER match.
 			headers = append(headers, &envoy_api_v2_route.HeaderMatcher{Name: hdr.Name,
 				HeaderMatchSpecifier: &envoy_api_v2_route.HeaderMatcher_ExactMatch{ExactMatch: ""},
 				InvertMatch:          true})
 		} else {
 			// Header presence and matching (literal) value needed.
-			headers = append(headers, &envoy_api_v2_route.HeaderMatcher{Name: hdr.Name,
-				HeaderMatchSpecifier: &envoy_api_v2_route.HeaderMatcher_ExactMatch{ExactMatch: value}})
+			if mismatch_action == cilium.HeaderMatch_FAIL_ON_MISMATCH {
+				if value != "" {
+					headers = append(headers, &envoy_api_v2_route.HeaderMatcher{Name: hdr.Name,
+						HeaderMatchSpecifier: &envoy_api_v2_route.HeaderMatcher_ExactMatch{ExactMatch: value}})
+				} else {
+					// Only header presence needed
+					headers = append(headers, &envoy_api_v2_route.HeaderMatcher{Name: hdr.Name,
+						HeaderMatchSpecifier: &envoy_api_v2_route.HeaderMatcher_PresentMatch{PresentMatch: true}})
+				}
+			} else {
+				log.Debugf("HeaderMatches: Adding %s: %s", hdr.Name, value)
+				headerMatches = append(headerMatches, &cilium.HeaderMatch{
+					MismatchAction: mismatch_action,
+					Name:           hdr.Name,
+					Value:          value,
+				})
+			}
 		}
 	}
 	if len(headers) == 0 {
 		headers = nil
 	} else {
 		SortHeaderMatchers(headers)
 	}
-	return headers
+	if len(headerMatches) == 0 {
+		headerMatches = nil
+	} else {
+		// SortHeaderMatches(headerMatches)
```

Comment:
it this commented by design?

Reply:
yes, we would optimally sort the headerMatches to avoid unnecessary updates due to different order, but that has not been implemented yet. Will also need to consider if this sorting is needed after all, as data is in a slice, so the order only changes is the order of rules in the imported policies changes? What do you think?

Reply:
Added this comment:
```
		// Optimally we should sort the headerMatches to avoid
		// updating the policy if only the order of the rules
		// has changed. Right now, when 'headerMatches' is a
		// slice (rather than a map) the order only changes if
		// the order of the rules in the imported policies
		// changes, so there is minimal likelihood of
		// unnecessary policy updates.
```